lib/trino-parquet/src/main/java/io/trino/parquet/reader/ChunkedInputStream.java

```diff
@@ -0,0 +1,125 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.parquet.reader;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Iterators;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.Slices;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.Objects;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class ChunkedInputStream
+        extends InputStream
+{
+    private final Iterator<SliceInput> chunks;
+    private SliceInput current;
+
+    public static ChunkedInputStream from(Iterator<Slice> slices)
+    {
+        return new ChunkedInputStream(Iterators.transform(slices, Slice::getInput));
+    }
+
+    @VisibleForTesting
+    ChunkedInputStream(Iterator<SliceInput> chunks)
+    {
+        this.chunks = requireNonNull(chunks, "chunks is null");
+        current = chunks.hasNext() ? chunks.next() : Slices.EMPTY_SLICE.getInput();
+    }
+
+    public Slice getSlice(int length)
+    {
+        if (length == 0) {
+            return Slices.EMPTY_SLICE;
+        }
+        while (current.available() == 0) {
+            checkArgument(chunks.hasNext(), "requested %s bytes but 0 was available", length);
+            current = chunks.next();
+        }
+        if (current.available() >= length) {
+            return current.readSlice(length);
+        }
+        // requested length crosses the slice boundary
+        byte[] bytes = new byte[length];
+        int bytesRead = read(bytes, 0, bytes.length);
+        checkArgument(bytesRead == length, "requested %s bytes but only %s was available", length, bytesRead);
+        return Slices.wrappedBuffer(bytes);
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len)
+    {
+        Objects.checkFromIndexSize(off, len, b.length);
+        if (len == 0) {
+            return 0;
+        }
+        while (current.available() == 0) {
+            if (!chunks.hasNext()) {
+                return -1;
+            }
+            current = chunks.next();
+        }
+
+        if (current.available() >= len) {
+            return current.read(b, off, len);
+        }
+
+        // requested length crosses the slice boundary
+        int leftToRead = len;
+        while (leftToRead > 0) {
+            int read = current.read(b, off, leftToRead);
+            if (read == -1) {
+                if (!chunks.hasNext()) {
+                    // no more data, return what has been read so far
+                    break;
+                }
+                current = chunks.next();
+            }
+            else {
+                leftToRead -= read;
+                off += read;
+            }
+        }
+        return len - leftToRead;
```

Comment:
Since we're allowed to return after reading fewer bytes than requested, we can simplify this part to just
```
return current.read(b, off, len);
```
and let the caller have the responsibility of doing while loop over this method.

Reply:
that is an excellent idea!